### ISSUE TITLE
mipchain is incomplete

### ISSUE THREAD

**Issue by promethe42** [dariomanesku/cmft#10]
Hello,

I'm trying to implement PBR in our open source x-platform engine minko (http://minko.io). So CMFT is a really amazing tool to work on cubemaps: thank you!

Still, I encountered a few issues: when I use the "radiance" filter, the mipchain is not complete.

* From a 2048x1536 hcross cubemap, I get a mipchain with the first level is missing or already blurred despite using `--excludeBase false`. The mipchain loads fine though, but the visual result is not as expected for roughness close to 0.
* From a 2048x1024 latlong texture, I get a 2048x2046 latlong output. The last line of that output mipmap chain is expected to be a 1x1 mipmap, but instead it's a 4x2 mipmap. As a result, the mipchain cannot be loaded.

Is this a bug or did I miss something?

Please advise,

**Comment by dariomanesku** [comment]
- _first level is already blurred despite using --excludeBase false_
You should use '--excludeBase true' insted.
--excludeBase true == exclude filter on first level cubemap. Exactly what you are asking for.

-  _From a 2048x1024 latlong texture, I get a 2048x2046 latlong output._
That doesn't sound right at all. Could you double check that and if that's really what is going on, please give me your input so I can reproduce the error and fix the problem.

- _The last line of that output mipmap chain is expected to be a 1x1 mipmap._
It's not, it depends on input parameters. Check out this screenshot from cmftStudio: ![cmftstudio_params](https://cloud.githubusercontent.com/assets/1139260/7006264/060a60e2-dc80-11e4-904c-9a55a06ef588.jpg)

You should have a look at: https://github.com/dariomanesku/cmftStudio
There you can easily experiment with input parameters and check out the results immediately.

**Comment by promethe42** [comment]
`--excludeBase true` does solve the 1st problem. Thank you!

What parameters will influence the number of mipmaps?

The main issue here is that for output formats that don't explicitly handle mipmaps (ie tga) you'll have to "parse" the mipmap chain out of the bitmap data. It's not a problem as long as you can actually compute the number of mipmaps based on the width/height of the bitmap AND the mipmap chain is actually complete. Correct?

**Comment by dariomanesku** [comment]
- _What parameters will influence the number of mipmaps?_
It's '--mipCount' or 'Num mipmap to filter' in the cmftStudio interface.
I think you can use '--mipCount 100', or some other big number, and that will make sure that all mipmap levels get filtered.

Usually, 1x1 and 2x2 mipmaps are practically useless. 4x4 is the lowest resolution you should use. With --mipCount you can discard the last two mipmap levels (as in the screenshot above).

Regarding tga, you can use something like this '--output0params tga,bgra8,facelist' and then you'll get:
![cmftstudio_output](https://cloud.githubusercontent.com/assets/1139260/7007753/be16be22-dc8c-11e4-98b0-8511173ca45c.jpg)
in format: name_face_miplevel_widthxheight.tga

I guess you could easily parse that, right?

**Comment by promethe42** [comment]
> I guess you could easily parse that, right?

Sure. But we're targeting HTML5 so we try to avoid having so many I/O.

> It's '--mipCount' or 'Num mipmap to filter' in the cmftStudio interface.

here is what I do:

```
$CONVERT_BIN $1 -resize 2048 $TMP_FILE.tga

$CMFT_BIN                                       \
    --input $TMP_FILE.tga                       \
    --filter radiance                           \
    --srcFaceSize 256                           \
    --edgeFixup warp                            \
    --glossScale 10                             \
    --mipCount 100                              \
    --excludeBase true                          \
    --glossBias 1                               \
    --lightingModel phongbrdf                   \
    --dstFaceSize 256                           \
    --numCpuProcessingThreads 4                 \
    --useOpenCL true                            \
    --clVendor anyGpuVendor                     \
    --deviceType gpu                            \
    --deviceIndex 0                             \
    --inputGammaNumerator 1.0                   \
    --inputGammaDenominator 1.0                 \
    --outputGammaNumerator 1.0                  \
    --outputGammaDenominator 1.0                \
    --outputNum 1                               \
    --output0 $TMP_DIR/$INPUT_FILENAME          \
    --output0params tga,bgra8,$OUTPUT_LAYOUT
```

I still have only 9 mipmaps in the output, with the final mipmap being 4x2.

Here is the source image: https://dl.dropboxusercontent.com/u/39247193/skybox_texture.jpg
Here is the result: https://dl.dropboxusercontent.com/u/39247193/skybox_texture-radiance.jpg

**Comment by dariomanesku** [comment]
Well, yes, that's the resolution of the last mipmap in latlong format. Have a look at the resolutions here:
![cmftstudio_sizes](https://cloud.githubusercontent.com/assets/1139260/7008241/fee19ec8-dc90-11e4-85d7-36318b17a4b2.jpg)

At level 0, you either have 6 cubemap faces of 256x256, or 1 latlong image of 1024x512.
In the same way: at level 8, you either have 6 cubemap faces of 1x1 or 1 latlong image of 4x2.

And, btw, mipmaped cubemap texture doesn't necessarily have to be 'complete' in the sense that it contains all mipmap levels down to 1x1. A cubemap texture with, for example, 3 mip map levels: 512x512, 256x256, 128x128 is perfectly valid.

**Comment by dariomanesku** [comment]
Regardless of this, I now notice some glitches in the base image that is converted to latlong. These appear in the image right where the edges of cubemap faces are. I'll have a look into it.

**Comment by promethe42** [comment]
> At level 0, you either have 6 cubemap faces of 256x256, or 1 latlong image of 1024x512.
In the same way: at level 8, you either have 6 cubemap faces of 1x1 or 1 latlong image of 4x2.

Why not 1x1 for the last level of the mipmap chain?
I know they might be useless. Still, it requires to add some extra flags to the mipchain parsing code to handle this special case. I'd rather not have to do this.
In the end, I'm actually quite happy to get all mipmaps in a separate file and merge them with `convert`. I guess if someone doesn't need those last levels he could simply not "merge" them at the final step.

Don't you think?

> Regardless of this, I now notice some glitches in the base image that is converted to latlong. These appear in the image right where the edges of cubemap faces are. I'll have a look into it.

Don't worry, I'm pretty sure the original cubemap is "broken".

**Comment by dariomanesku** [comment]
Latlong image has aspect ratio of 2:1, and the lowest latlong that could represent an environment is 4x2. Lower than that it would be:
- 2x1 == total 2 pixels, not enough for the entire environment. In order to represent a minimal cubemap, 6 pixels are required. 
- 1x1 == invalid latlong format!

It's should not be a special case for mipmap chain parsing code. If it is, make a placeholder image of 2x1/1x1 (or whatever that mipmap parsing code requires) and use that along with output from cmft. Either way, those images will be discarded at some point because the last cubemap mipmap image will be obtained from the 4x2 latlong image.


- _I'm actually quite happy to get all mipmaps in a separate file and merge them with convert._

Could you merge facelist output with `convert`? If that's possible, that's the best way to go.

**Comment by promethe42** [comment]
> This should not be a special case for mipmap chain parsing code.

I don't know how to do it any other way: if the mipmap chain is not complete, I can't parse it without knowing the exact number of levels. That's not practical, since changing the asset would require to change the loading code.

> make a placeholder image of 2x1/1x1

That's what I'm doing now using `convert`. I just wanted to talk about this with you.
Maybe we might add a "--keepInvalidMipMaps" option to handle this case?

> Could you merge facelist output with convert? If that's possible, that's the best way to go.

It's not in my case since I can't do cube map sampling with LOD in WebGL/OpenGL ES 2.
I have to use latlong. Using cmft simply to convert to the latlong layout is actually another great feature!

**Comment by promethe42** [comment]
Here is how I generate the missing mipmaps using `convert`:

```
TMP_DIR=$(mktemp -d)
TMP_FILE=$(mktemp)

$CONVERT_BIN $1 -resize 2048 $TMP_FILE.tga

$CMFT_BIN                                       \
    --input $TMP_FILE.tga                       \
    --filter radiance                           \
    --srcFaceSize 256                           \
    --edgeFixup warp                            \
    --glossScale 10                             \
    --mipCount 100                              \
    --excludeBase true                          \
    --glossBias 1                               \
    --lightingModel phongbrdf                   \
    --dstFaceSize 256                           \
    --numCpuProcessingThreads 4                 \
    --useOpenCL true                            \
    --clVendor anyGpuVendor                     \
    --deviceType gpu                            \
    --deviceIndex 0                             \
    --inputGammaNumerator 1.0                   \
    --inputGammaDenominator 1.0                 \
    --outputGammaNumerator 1.0                  \
    --outputGammaDenominator 1.0                \
    --outputNum 1                               \
    --output0 $TMP_DIR/$INPUT_FILENAME          \
    --output0params tga,bgra8,$OUTPUT_LAYOUT

if [ $OUTPUT_LAYOUT == "latlong" ]; then
    $CONVERT_BIN $TMP_DIR/${INPUT_FILENAME}_8_4x2.tga -resize 2x1 $TMP_DIR/${INPUT_FILENAME}_9_2x1.tga
    $CONVERT_BIN $TMP_DIR/${INPUT_FILENAME}_8_4x2.tga -resize 1x1 $TMP_DIR/${INPUT_FILENAME}_10_1x1.tga
fi

# $CONVERT_BIN $TMP_FILE.tga $TMP_DIR/* -background black -append $OUTPUT_DIR/$INPUT_FILENAME-radiance.$OUTPUT_EXTENSION
TMP_FILES=`ls $TMP_DIR | sed s/${INPUT_FILENAME}_// | sort -g | sed s#^#$TMP_DIR/${INPUT_FILENAME}_#`

$CONVERT_BIN $TMP_FILES -background black -append $OUTPUT_DIR/$INPUT_FILENAME-radiance.$OUTPUT_EXTENSION
```

**Comment by promethe42** [comment]
Actually it doesn't solve nothing since mipmaps with invalid ratio will make the mipmap chain just as hard to parse... "measuring" the width of the last mipmap might help, but it's not a very versatile situation.

**Comment by dariomanesku** [comment]
You could make this: https://dl.dropboxusercontent.com/u/39247193/skybox_texture-radiance.jpg
with cubemap faces instead of latlong. That way, the aspect ratio will be correct for the parser.
Then, you either live with 6 images like that or you merge them vertically in a single image and calculate their offset in the code.

If you have a suggestion on how to make this whole process less painful, I would like to know.

**Comment by promethe42** [comment]
> with cubemap faces instead of latlong

Cubemap faces will be sampled using a samplerCube/textureCube. And I can't use LOD with those.
AFAIK latlong is the only viable layout since it won't be an OpenGL cube texture.

I think I have no choice and I'll have to add some kind of "layout" flag when loading the assets (I'll have to find another name though, since we already use the word "layout" for our scene graph).

I'll keep you posted when I settle for an implementation and how the final result looks like.
